sdk/iot/hub/samples/methods/src/iot_hub_methods_example.c

```diff
@@ -0,0 +1,430 @@
+      if ((result = send_method_response(&method_request, 404, 0xFFFF)) != AZ_OK)
+      {
+        printf("Unable to send %d response, status %d", 404, result);
+        return result;
+      }
+    }
+  }
+
+  MQTTClient_freeMessage(&message);
+  MQTTClient_free(topicName);
+
+  return 1;
+}
+
+static int connect_device()
+{
+  int rc;
+
+  MQTTClient_SSLOptions mqtt_ssl_options = MQTTClient_SSLOptions_initializer;
+  MQTTClient_connectOptions mqtt_connect_options = MQTTClient_connectOptions_initializer;
+
+  char username[128];
+  size_t username_length;
+  if ((rc = az_iot_hub_client_get_user_name(&client, username, sizeof(username), &username_length))
+      != AZ_OK)
+
+  {
+    printf("Failed to get MQTT clientId, return code %d\n", rc);
+    return rc;
+  }
+
+  mqtt_connect_options.username = username;
+  mqtt_connect_options.password = NULL;
+
+  mqtt_ssl_options.keyStore = (char*)x509_cert_pem_file;
+  if (*x509_trust_pem_file != '\0')
+  {
+    mqtt_ssl_options.trustStore = (char*)x509_trust_pem_file;
+  }
+
+  mqtt_connect_options.ssl = &mqtt_ssl_options;
+
+  if ((rc = MQTTClient_connect(mqtt_client, &mqtt_connect_options)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to connect, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+static int subscribe()
+{
+  int rc;
+
+  char c2d_topic[128];
+  size_t c2d_topic_length;
+  if ((rc = az_iot_hub_client_methods_get_subscribe_topic_filter(
+           &client, c2d_topic, sizeof(c2d_topic), &c2d_topic_length))
+      != AZ_OK)
+
+  {
+    printf("Failed to get C2D MQTT SUB topic filter, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_subscribe(mqtt_client, c2d_topic, 1)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to subscribe, return code %d\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+int main()
+{
+  int rc;
+
+  if ((rc = read_configuration_and_init_client()) != AZ_OK)
+  {
+    printf("Failed to read configuration from environment variables, return code %d\n", rc);
+    return rc;
+  }
+
+  char client_id[128];
+  size_t client_id_length;
+  if ((rc
+       = az_iot_hub_client_get_client_id(&client, client_id, sizeof(client_id), &client_id_length))
+      != AZ_OK)
+  {
+    printf("Failed to get MQTT clientId, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_create(
+           &mqtt_client, mqtt_endpoint, client_id, MQTTCLIENT_PERSISTENCE_NONE, NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to create MQTT client, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = MQTTClient_setCallbacks(mqtt_client, NULL, NULL, on_received, NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to set MQTT callbacks, return code %d\n", rc);
+    return rc;
+  }
+
+  if ((rc = connect_device()) != 0)
+  {
+    return rc;
+  }
+
+  if ((rc = subscribe()) != 0)
+  {
+    return rc;
+  }
+
+  printf("Subscribed to topics.\n");
+
+  printf("Waiting for activity. [Press ENTER to abort]\n");
```

Comment:
See other comment: this will likely timeout at MQTT-protocol level. We should instead convert to sync MQTT_recv and use timeouts instead of getchar.

Reply:
backlog ✔️